api/scenario_api/src/scenario_api_node.cpp

```diff
@@ -17,228 +17,247 @@
 #include <scenario_api/scenario_api_core.h>
 #include <scenario_api_utils/scenario_api_utils.h>
 
+namespace {
+// TODO(fred-apex-ai) how to print msg to stdout? https://answers.ros.org/question/365776/how-to-print-a-message-to-stdout-in-ros2/
+// Once available, use the `to_yaml` function instead instead of this handcrafted operator
+std::ostream & operator<<(std::ostream & out, const geometry_msgs::msg::Pose &pose) {
+  out << "position = {x=" << pose.position.x << ", y=" << pose.position.y << ", z="
+  << pose.position.z << "}" << std::endl;
+  out << "orientation = {x=" << pose.orientation.x << ", y=" << pose.orientation.y << ", z="
+  << pose.orientation.z << ", w=" << pose.orientation.w << "}" << std::endl;
+  return out;
+}
+}
+
 int main(int argc, char ** argv)
 {
-  ros::init(argc, argv, "scenario_api");
-  ScenarioAPI obj;
+  rclcpp::init(argc, argv);
+  // TODO(fred-apex-ai) Throws std::bad_weak_ptr when launched via scenario_api_sample.launch.xml
```

Comment:
Has this been fixed?

Reply:
I am looking into it, here the backtrace
```
#4  0x00007f8abf7884e7 in std::terminate() () from /lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007f8abf788799 in __cxa_throw () from /lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007f8abfc417d2 in std::__throw_bad_weak_ptr () at /usr/include/c++/9/bits/shared_ptr_base.h:84
#7  0x00007f8abfc44c10 in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count (this=0x7ffc400fadc8, __r=...) at /usr/include/c++/9/bits/shared_ptr_base.h:900
#8  0x00007f8abfc44685 in std::__shared_ptr<rclcpp::Node, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<rclcpp::Node, void> (this=0x7ffc400fadc0, __r=...) at /usr/include/c++/9/bits/shared_ptr_base.h:1193
#9  0x00007f8abfc43ce3 in std::shared_ptr<rclcpp::Node>::shared_ptr<rclcpp::Node, void> (this=0x7ffc400fadc0, __r=std::weak_ptr<class rclcpp::Node> (empty) = {...}) at /usr/include/c++/9/bits/shared_ptr.h:276
#10 0x00007f8abfc42d9f in std::enable_shared_from_this<rclcpp::Node>::shared_from_this (this=0x556188e38f38) at /usr/include/c++/9/bits/shared_ptr.h:651
#11 0x00007f8abfc3ee41 in ScenarioAPI::ScenarioAPI (this=0x556188e38f30) at /home/frederik.beaujean/AutowareArchitectureProposal/src/simulator/scenario_runner/api/scenario_api/src/scenario_api_core.cpp:22
#12 0x00005561871a525d in __gnu_cxx::new_allocator<ScenarioAPI>::construct<ScenarioAPI> (this=0x7ffc400fb08f, __p=0x556188e38f30) at /usr/include/c++/9/ext/new_allocator.h:147
#13 0x00005561871a5029 in std::allocator_traits<std::allocator<ScenarioAPI> >::construct<ScenarioAPI> (__a=..., __p=0x556188e38f30) at /usr/include/c++/9/bits/alloc_traits.h:484
#14 0x00005561871a4d5c in std::_Sp_counted_ptr_inplace<ScenarioAPI, std::allocator<ScenarioAPI>, (__gnu_cxx::_Lock_policy)2>::_Sp_counted_ptr_inplace<>(std::allocator<ScenarioAPI>) (this=0x556188e38f20, __a=...) at /usr/include/c++/9/bits/shared_ptr_base.h:548
#15 0x00005561871a4854 in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<ScenarioAPI, std::allocator<ScenarioAPI>>(ScenarioAPI*&, std::_Sp_alloc_shared_tag<std::allocator<ScenarioAPI> >) (this=0x7ffc400fb278, __p=@0x7ffc400fb270: 0x0, __a=...) at /usr/include/c++/9/bits/shared_ptr_base.h:679
#16 0x00005561871a455c in std::__shared_ptr<ScenarioAPI, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<std::allocator<ScenarioAPI>>(std::_Sp_alloc_shared_tag<std::allocator<ScenarioAPI> >) (this=0x7ffc400fb270, __tag=...) at /usr/include/c++/9/bits/shared_ptr_base.h:1344
#17 0x00005561871a407f in std::shared_ptr<ScenarioAPI>::shared_ptr<std::allocator<ScenarioAPI>>(std::_Sp_alloc_shared_tag<std::allocator<ScenarioAPI> >) (this=0x7ffc400fb270, __tag=...) at /usr/include/c++/9/bits/shared_ptr.h:359
#18 0x00005561871a3720 in std::allocate_shared<ScenarioAPI, std::allocator<ScenarioAPI>>(std::allocator<ScenarioAPI> const&) (__a=...) at /usr/include/c++/9/bits/shared_ptr.h:702
#19 0x00005561871a2a9f in std::make_shared<ScenarioAPI> () at /usr/include/c++/9/bits/shared_ptr.h:718
#20 0x00005561871972f7 in main (argc=46, argv=0x7ffc400fc308) at /home/frederik.beaujean/AutowareArchitectureProposal/src/simulator/scenario_runner/api/scenario_api/src/scenario_api_node.cpp:37

```

Reply:
haha I just read this paragraph this week and missed the part in parentheses. That sucks, we have to create other objects after the ctor. I don't like this passing a node ptr thing

> 
> It is permitted to call shared_from_this only on a previously shared object, i.e. on an object managed by std::shared_ptr (in particular, shared_from_this cannot be called during construction of *this).
> 
> Otherwise the behavior is undefined (until C++17)std::bad_weak_ptr is thrown (by the shared_ptr constructor from a default-constructed weak_this) (since C++17). 

This is the culprit

```c++
ScenarioAPI::ScenarioAPI()
: Node("scenario_api"),
  simulator_api_{std::make_shared<ScenarioAPISimulator>(Node::shared_from_this())},
```

Reply:
fixed in d51a9a6